cloud/pkg/nodeupgradejobcontroller/controller/downstream.go

```diff
@@ -162,65 +178,121 @@ func (dc *DownstreamController) nodeUpgradeJobAdded(upgrade *v1alpha1.NodeUpgrad
 
-		upgradeReq := commontypes.NodeUpgradeJobRequest{
-			UpgradeID:   upgrade.Name,
-			HistoryID:   uuid.New().String(),
-			UpgradeTool: upgrade.Spec.UpgradeTool,
-			Version:     upgrade.Spec.Version,
-			Image:       image,
-		}
+	upgradeReq := commontypes.NodeUpgradeJobRequest{
+		UpgradeID:   upgrade.Name,
+		HistoryID:   uuid.New().String(),
+		UpgradeTool: upgrade.Spec.UpgradeTool,
+		Version:     upgrade.Spec.Version,
+		Image:       image,
+	}
+
+	msg.BuildRouter(modules.NodeUpgradeJobControllerModuleName, modules.NodeUpgradeJobControllerModuleGroup, resource, NodeUpgrade).
+		FillBody(upgradeReq)
+
+	err = dc.messageLayer.Send(*msg)
+	if err != nil {
+		klog.Errorf("Failed to send upgrade message %v due to error %v", msg.GetID(), err)
+		return
+	}
+
+	// process time out: cloud did not receive upgrade feedback from edge
+	// send upgrade timeout response message to upstream
+	go dc.handleNodeUpgradeJobTimeout(node, upgrade.Name, upgrade.Spec.Version, upgradeReq.HistoryID, upgrade.Spec.TimeoutSeconds)
+
+	// mark Upgrade state upgrading
+	status := &v1alpha1.UpgradeStatus{
+		NodeName: node,
+		State:    v1alpha1.Upgrading,
+		History: v1alpha1.History{
+			HistoryID:   upgradeReq.HistoryID,
+			UpgradeTime: time.Now().String(),
+		},
+	}
+	err = patchNodeUpgradeJobStatus(dc.crdClient, upgrade, status)
+	if err != nil {
+		// not return, continue to mark node unschedulable
+		klog.Errorf("Failed to mark Upgrade upgrading status: %v", err)
+	}
+
+	// mark edge node unschedulable
+	// the effect is like running cmd: kubectl drain <node-to-drain> --ignore-daemonsets
+	unscheduleNode := v1.Node{}
+	unscheduleNode.Spec.Unschedulable = true
 
-		msg.BuildRouter(modules.NodeUpgradeJobControllerModuleName, modules.NodeUpgradeJobControllerModuleGroup, resource, NodeUpgrade).
-			FillBody(upgradeReq)
+	// add a upgrade label
+	unscheduleNode.Labels = map[string]string{NodeUpgradeJobStatusKey: NodeUpgradeJobStatusValue}
+	byteNode, err := json.Marshal(unscheduleNode)
+	if err != nil {
+		klog.Warningf("marshal data failed: %v", err)
+		return
+	}
 
-		err := dc.messageLayer.Send(*msg)
+	_, err = dc.kubeClient.CoreV1().Nodes().Patch(context.Background(), node, apimachineryType.StrategicMergePatchType, byteNode, metav1.PatchOptions{})
+	if err != nil {
+		klog.Errorf("failed to drain node %s: %v", node, err)
+		return
+	}
+}
+
+// selectConcurrentNodes select the nodes to do upgrade operation, and put it into channel nodesChan
+func (dc *DownstreamController) selectConcurrentNodes(nodesChan chan string, allNodes []string, upgrade *v1alpha1.NodeUpgradeJob) {
+	// the default concurrency is 1
+	// this means that we will upgrade nodes one by one
+	// only when the last one node upgrade finished, we'll continue to upgrade the next one node
+	concurrency := 1
+	if upgrade.Spec.Concurrency != 0 {
+		concurrency = int(upgrade.Spec.Concurrency)
+	}
+
+	timeout := int(*upgrade.Spec.TimeoutSeconds) * (len(allNodes) + 1) / concurrency
```

Comment:
Can you explain why we calculate timeout like this?

Reply:
use this just for ensure we can process all the timeout circumstance.